R/class-yarn.R

```diff
@@ -0,0 +1,134 @@
+    #' ex1 <- tinkr::yarn$new(path)
+    #' ex1
+    #' path2 <- system.file("extdata", "example2.Rmd", package = "tinkr")
+    #' ex2 <- tinkr::yarn$new(path2)
+    #' ex2
+    initialize = function(path = NULL, encoding = "UTF-8", sourcepos = FALSE) {
+      if (is.null(path)) {
+        xml <- list(yaml = NULL, body = NULL)
+      } else {
+        xml <- to_xml(path, encoding, sourcepos)
+      }
+
+      self$path <- path
+      self$yaml <- xml$yaml
+      self$body <- xml$body
+      self$ns   <- xml2::xml_ns_rename(xml2::xml_ns(xml$body), d1 = "md")
+      invisible(self)
+    },
+
+    #' @description reset a yarn document from the original file
+    #' @examples
+    #'
+    #' path <- system.file("extdata", "example1.md", package = "tinkr")
+    #' ex1 <- tinkr::yarn$new(path)
+    #' # OH NO
+    #' ex1$body
+    #' ex1$body <- xml2::xml_missing()
+    #' ex1$reset()
+    #' ex1$body
+    reset = function() {
+      x <- to_xml(self$path)
+      self$body <- x$body
+      self$yaml <- x$yaml
+      invisible(self)
+    },
+
+    #' @description Write a yarn document to markdown/Rmarkdown
+    #'
+    #' @param path path to the file you want to write
+    #' @param stylesheet_path path to the xsl stylesheet to convert XML to markdown.
+    #' @examples
+    #' path <- system.file("extdata", "example1.md", package = "tinkr")
+    #' ex1 <- tinkr::yarn$new(path)
+    #' ex1
+    #' tmp <- tempfile()
+    #' try(readLines(tmp)) # nothing in the file
+    #' ex1$write(tmp)
+    #' head(readLines(tmp)) # now a markdown file
+    #' unlink(tmp)
+    write = function(path = NULL, 
+      stylesheet_path = system.file("extdata", "xml2md_gfm.xsl", package = "tinkr")){
+      if (is.null(path)) {
+        stop("Please provide a file path", call. = FALSE)
+      }
+      to_md(self, path, stylesheet_path)
+      invisible(self)
+    },
+    
+    #' @description add an arbitrary markdown element to the document
+    #'
+    #' @param md a string of markdown formatted text.
+    #' @param where the location in the document to add your markdown text.
+    #'   This is passed on to [xml2::xml_add_child()]. Defaults to 0, which
+    #'   indicates the very top of the document.
+    #' @examples
+    #' path <- system.file("extdata", "example2.Rmd", package = "tinkr")
+    #' ex <- tinkr::yarn$new(path)
+    #' # two headings, no lists
+    #' xml2::xml_find_all(ex$body, "md:heading", ex$ns)
+    #' xml2::xml_find_all(ex$body, "md:list", ex$ns)
+    #' ex$add_md(
+    #'   "# Hello\n\nThis is *new* formatted text from `{tinkr}`!",
+    #'   where = 1L
+    #' )$add_md(
+    #'   " - This\n - is\n - a new list",
+    #'   where = 2L
+    #' )
+    #' # three headings
+    #' xml2::xml_find_all(ex$body, "md:heading", ex$ns)
+    #' xml2::xml_find_all(ex$body, "md:list", ex$ns)
+    #' tmp <- tempfile()
+    #' ex$write(tmp)
+    #' readLines(tmp, n = 20) 
+    add_md = function(md, where = 0L) {
```

Comment:
To keep with the knitting metaphor, we could call this `join_md()` instead, but I think it may be confusing with SQL joins.